1-Draft/RFC00XX-Native-Command-Error-Handling.md

```diff
@@ -0,0 +1,140 @@
+---
+RFC: RFC00XX
+Author: Jason Helmick
+Status: Draft
+Area: Core
+Comments Due: 10/31/2020
+---
+
+# Native Command Error Handling
+
+PowerShell scripts using native commands would benefit from being able to use error handling
+features like those used by cmdlets.
+
+## Motivation
+
+In PowerShell by default, script processing continues when non-terminating errors occur. This is a
+benefit when expecting non-terminating errors in normal execution such as non-responsive computers
+from a list. This default behavior is controlled with the preference variable
+`$ErrorActionPreference` default of `Continue`.
+
+In production, often customers prefer that script execution stops when a non-terminating error
+occurs. This is particularly true in CI where the preference is to fail fast. PowerShell currently
+supports customers with this ability by setting the `$ErrorActionPreference` variable in the script
+to `Stop`.
+
+```Powershell
+$ErrorActionPreference = 'Stop'
+```
+
+Native commands usually return an exit code to the calling application which will be zero for
+success or non-zero for failure. However, native commands currently do not participate in the
+PowerShell error stream. Redirected `stderr` output is not interpreted the same as the PowerShell
+error stream as many native commands use `stderr` as information/verbose stream and thus only the
+exit code matters. Users working with native commands in their scripts will need to check the
+execution status after each call using a helper function similar to below:
+
+```Powershell
+if ($LASTEXITCODE -ne 0)
+{
+    throw "Command failed. See above errors for details"
+}
+ ```
+
+Simply relaying the errors through the error stream isn't the solution. The example itself doesn't
+support all cases as `$?` can be false from a cmdlet or function error, making `$LASTEXITCODE`
+stale.
+
+In POSIX shells, this need to terminate on command error is addressed by the `set -e` configuration,
```

Comment:
I think it is implied by the current revision, but let me spell it out to see if my understanding is correct:

* _Unlike_ `bash`'s `set -e`, **PowerShell will _not_ (and should not) make the failure detection dependent on whether or not it is used in a _conditional_** (e.g., of an `if` statement).

   * Unlike in POSIX-like shells, where conditionals act on _exit codes_, PowerShell's conditionals act on _data_, so it make sense to treat failure detection separately and to thereby  avoid the rule complexity surrounding `set -e`.

* _Like_ `bash`'s `set -e`, **PowerShell _will_ bypass failure detection if a failure occurs in a non-final segment of a _pipeline chain_**, so as to allow overriding `$PSNativeCommandErrorAction` _ad hoc_.

Examples:

```powershell
$PSNativeCommandErrorAction = 'Stop'

# SHOULD fail, if `ls` reports a nonzero exit code.
if (-not ($name = /bin/ls nosuch.txt))  {
  $name = 'default.txt'
}

# Should NOT fail, because || is used to explicitly handle the failure:
/bin/ls nosuch.txt || 'default.txt'
```